File: julia/src/weights.jl
# Integration weights, computed from the connectivity.
#
# Section 3 says a weight array is "computed from connectivity, never
# imported", and it names no formula, so this file is the formula:
# the measure of a cell from its type and its nodes, and for a node
# the lumped share of the cells it belongs to.  Nothing here is per
# file or per family; a cell type is measured for every file or it is
# refused for every file, and the refusal says which code it met.

"""The topological dimension of each cell type of section 20, which is
the power of length its measure has."""
const CELL_DIM = Dict{Int,Int}(1 => 0, 3 => 1, 5 => 2, 7 => 2, 9 => 2,
                               10 => 3, 12 => 3, 13 => 3, 14 => 3,
                               21 => 1, 22 => 2, 23 => 2, 24 => 3,
                               25 => 3, 26 => 3, 27 => 3)

"""The cell types this package measures: the straight-sided ones.  A
quadratic cell's measure is an integral over its curved geometry,
which this version does not compute and will not approximate
silently."""
const MEASURED_CELLS = (3, 5, 7, 9, 10, 12, 13, 14)

# The straight-sided decompositions, in a cell's own node order
# (section 20 keeps VTK's): a polygon and a quadrilateral fan into
# triangles from their first node, a hexahedron into five tetrahedra,
# a wedge into three, a pyramid into two.  Each is exact for the
# planar-faced cell.
const HEX_TETS = ((1, 2, 3, 6), (1, 3, 6, 8), (1, 3, 4, 8),
                  (1, 5, 6, 8), (3, 6, 7, 8))
const WEDGE_TETS = ((1, 2, 3, 4), (2, 3, 4, 5), (3, 4, 5, 6))
const PYRAMID_TETS = ((1, 2, 3, 5), (1, 3, 4, 5))

sub3(a, b) = (a[1] - b[1], a[2] - b[2], a[3] - b[3])
cross3(a, b) = (a[2] * b[3] - a[3] * b[2], a[3] * b[1] - a[1] * b[3],
                a[1] * b[2] - a[2] * b[1])
dot3(a, b) = a[1] * b[1] + a[2] * b[2] + a[3] * b[3]
norm3(a) = sqrt(a[1]^2 + a[2]^2 + a[3]^2)

triangle_area(a, b, c) = 0.5 * norm3(cross3(sub3(b, a), sub3(c, a)))
tetra_volume(a, b, c, d) =
    abs(dot3(sub3(b, a), cross3(sub3(c, a), sub3(d, a)))) / 6

"""One node of one instance as a three-vector, whatever the support's
spatial dimension."""
@inline function point3(coords::AbstractArray{Float64,3}, inst::Int, i::Int)
    d = size(coords, 3)
    return (coords[inst, i, 1],
            d >= 2 ? coords[inst, i, 2] : 0.0,
            d >= 3 ? coords[inst, i, 3] : 0.0)
end

"""The measure of one cell: its length, its area or its volume,
according to its type."""
function cell_measure(code::Int, pts, path)
    code == 3 && return norm3(sub3(pts[2], pts[1]))
    if code == 5 || code == 7 || code == 9
        a = 0.0
        for i in 2:(length(pts) - 1)
            a += triangle_area(pts[1], pts[i], pts[i + 1])
        end
        return a
    end
    code == 10 && return tetra_volume(pts[1], pts[2], pts[3], pts[4])
    tets = code == 12 ? HEX_TETS : code == 13 ? WEDGE_TETS :
           code == 14 ? PYRAMID_TETS : nothing
    tets === nothing && throw(unmeasured(code, path))
    return sum(tetra_volume(pts[t[1]], pts[t[2]], pts[t[3]], pts[t[4]])
               for t in tets)
end

unmeasured(code::Int, path) = MestraError("E21", path,
    haskey(CELL_DIM, code) ?
    "this package does not measure cell type $(code); it measures " *
    "the straight-sided types " * join(MEASURED_CELLS, ", ") *
    ", and a quadratic cell's measure is an integral over its curved " *
    "geometry that it will not approximate silently" :
    "cell type $(code) is not one of section 20")

"""The dimension every cell of a support has, which is the dimension
its measure has.  Cells of two dimensions in one support have no one
measure, so that is refused rather than added up."""
function cell_dimension(types::Vector{UInt8}, path)
    dims = Set{Int}()
    for t in types
        code = Int(t)
        code in MEASURED_CELLS || throw(unmeasured(code, path))
        push!(dims, CELL_DIM[code])
    end
    length(dims) <= 1 || throw(MestraError(nothing, path,
        "this support has cells of dimension " *
        join(sort(collect(dims)), " and ") * ", which have no one " *
        "measure between them; weigh a support whose cells are all of " *
        "one dimension"))
    return isempty(dims) ? 0 : first(dims)
end

"""The coordinates as (instance, node, component), with one instance
when they do not vary, and what they vary along."""
function coordinate_instances(ds::Dataset, sup::Support)
    c = sup.coordinates
    c === nothing && throw(MestraError("E03", "/supports/$(sup.name)",
        "this support carries no coordinates, so there is nothing to " *
        "measure"))
    p = permute_to_logical(ds, c)
    a = Array{Float64}(collect(parent(p)))
    first(dimnames(p)) in (:node, :cell) &&
        (a = reshape(a, 1, size(a)...))
    return (a, c.varies === nothing ? "none" : c.varies)
end

"""The lumped share of the intervals between neighbouring nodes of an
axis support, which is the same rule as the mesh one with the
intervals as the cells.  The nodes are taken in coordinate order,
which is the only connectivity an axis has."""
function axis_node_weights(x::AbstractVector{Float64})
    n = length(x)
    n == 0 && return Float64[]
    n == 1 && return [0.0]
    ord = sortperm(x)
    w = zeros(Float64, n)
    for k in 1:(n - 1)
        h = (x[ord[k + 1]] - x[ord[k]]) / 2
        w[ord[k]] += h
        w[ord[k + 1]] += h
    end
    return w
end

"""The units of a length raised to a power, or nothing for a unit
string this cannot raise, since a weight needs no units (E11)."""
function unit_power(u::Union{Nothing,AbstractString}, d::Int)
    u === nothing && return nothing
    s = String(strip(String(u)))
    isempty(s) && return nothing
    (d == 1 || s == "1") && return s
    (occursin(' ', s) || occursin('-', s) || occursin('/', s) ||
     occursin('^', s) || isdigit(last(s))) && return nothing
    return s * string(d)
end

"""
    compute_weights(ds, support, location) -> Matrix{Float64}

The weights themselves, as (instance, site): the measure of each cell
for `:cell`, and for `:node` the lumped share of the measure of the
cells each node belongs to.  There is one instance unless the
coordinates vary, in which case there is one per row or per group
category, because a mesh that moves has a measure that moves with it.

`compute_weights!` is the one to call; this is what it computes.
"""
function compute_weights(ds::Dataset, sup::Support, location::Symbol)
    return weight_matrix(ds, sup, location)[1]
end

compute_weights(ds::Dataset, name::AbstractString, location::Symbol) =
    compute_weights(ds, named_support(ds, name), location)

function weight_matrix(ds::Dataset, sup::Support, location::Symbol)
    path = "/supports/$(sup.name)"
    location in (:node, :cell) || throw(MestraError(nothing, path,
        "weights live at the nodes or at the cells; pass `location` as " *
        "`:node` or `:cell`"))
    coords, varies = coordinate_instances(ds, sup)
    ninst, nnode, _ = size(coords)
    if sup.kind == "axis"
        location === :cell && throw(MestraError("E38", path,
            "a support of kind axis has no cells; ask for `:node` weights"))
        w = zeros(Float64, ninst, nnode)
        for i in 1:ninst
            w[i, :] = axis_node_weights(Float64[coords[i, n, 1]
                                                for n in 1:nnode])
        end
        return (w, varies, unit_power(sup.coordinates.units, 1))
    end
    sup.kind == "mesh" || throw(MestraError(nothing, path,
        "a support of kind $(sup.kind) has no measure"))
    types, offsets, conn =
        sup.cell_types, sup.cell_offsets, sup.cell_connectivity
    (types === nothing || offsets === nothing || conn === nothing) &&
        throw(MestraError("E41", path,
            "this support's cells were not read, so there is no " *
            "connectivity to measure"))
    dim = cell_dimension(types, path * "/cell_types")
    ncell = length(types)
    out = zeros(Float64, ninst, location === :cell ? ncell : nnode)
    for j in 1:ncell
        nodes = Int[conn[k] + 1 for k in (offsets[j] + 1):offsets[j + 1]]
        for i in 1:ninst
            m = cell_measure(Int(types[j]),
                             [point3(coords, i, n) for n in nodes],
                             path * "/cell_types")
            if location === :cell
                out[i, j] = m
            else
                share = m / length(nodes)
                for n in nodes
                    out[i, n] += share
                end
            end
        end
    end
    return (out, varies, unit_power(sup.coordinates.units, dim))
end

"""
    compute_weights!(ds, support, location; name = "weight", units)

Add the integration weights of a support to the dataset, at the nodes
or at the cells, and give back the slot.

The array has role `weight`, is named `weight` at both locations
unless you name it, carries the units of the coordinates raised to the
support's dimension, and is marked `recomputed`, which is what W06
asks for.  It varies the way the coordinates vary: a fixed mesh has
one instance, a parametric family one per group category, a moving
mesh one per row.

    s = Mestra.add_mesh_support!(ds, "s0"; ...)
    Mestra.compute_weights!(ds, s, :node)
    Mestra.integrate(ds, "pressure")       # uses it

Section 3 says a weight is computed from the connectivity and never
imported, so this is the only way one gets into a file.
"""
function compute_weights!(ds::Dataset, sup::Support, location::Symbol;
                          name::AbstractString = "weight", units = nothing)
    w, varies, u = weight_matrix(ds, sup, location)
    units === nothing && (units = u)
    axis = location === :cell ? :cell : :node
    data = varies == "none" ? vec(w) : w
    dims = varies == "none" ? (axis,) :
           (Symbol(varies == "row" ? "row" : varies), axis)
    add = location === :cell ? add_cell_array! : add_node_array!
    return add(ds, sup, name, data; role = :weight, units = units,
               dims = dims, recomputed = true)
end

compute_weights!(ds::Dataset, name::AbstractString, location::Symbol;
                 kwargs...) =
    compute_weights!(ds, named_support(ds, name), location; kwargs...)

"""The support called `name`, or a refusal that lists the ones there
are."""
function named_support(ds::Dataset, name::AbstractString)
    i = support_by_name(ds, String(name))
    i === nothing && throw(MestraError(nothing, "/supports",
        "no support called `$(name)`; this file has " *
        (isempty(ds.supports) ? "none" :
         join(["`" * s.name * "`" for s in support_order(ds)], ", "))))
    return ds.supports[i]
end

named_support(::Dataset, ::Nothing) = throw(MestraError("E39", "/",
    "this slot names no support, so it has no nodes or cells to work over"))

File: cpp/src/weights.cpp
// Cell measures and the weights that follow from them.  Nothing here
// reads or writes a file: it works on a Support that is already in
// memory, which is the only place connectivity and coordinates are
// both to hand.
#include "mestra/weights.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <cstddef>

#include "mestra/io.hpp"
#include "names.hpp"

namespace mestra {
namespace {

[[noreturn]] void refuse(const std::string& path,
                         const std::string& message) {
  throw Error("", "mestra::compute_weights " + path + ": " + message);
}

// A point of the flat (node, component) block, padded to three
// dimensions so that one set of formulas covers d = 1, 2 and 3.
struct Point {
  double x = 0.0;
  double y = 0.0;
  double z = 0.0;
};

Point point_at(const std::vector<double>& coordinates,
               std::int64_t components, std::int64_t node) {
  Point p;
  const std::size_t at =
      static_cast<std::size_t>(node) * static_cast<std::size_t>(components);
  if (at < coordinates.size()) p.x = coordinates[at];
  if (components > 1 && at + 1 < coordinates.size()) p.y = coordinates[at + 1];
  if (components > 2 && at + 2 < coordinates.size()) p.z = coordinates[at + 2];
  return p;
}

Point minus(const Point& a, const Point& b) {
  return Point{a.x - b.x, a.y - b.y, a.z - b.z};
}

Point cross(const Point& a, const Point& b) {
  return Point{a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z,
               a.x * b.y - a.y * b.x};
}

double dot(const Point& a, const Point& b) {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

double length(const Point& a) { return std::sqrt(dot(a, a)); }

// The area of a planar polygon, by Newell's method.  It is the
// shoelace formula in two dimensions and the projected area in three,
// it is exact for any simple polygon rather than only a convex one,
// and it needs no triangulation to choose, so a triangle, a
// quadrilateral and a polygon of any size are one case here.
double polygon_area(const std::vector<Point>& p) {
  Point n;
  for (std::size_t i = 0; i < p.size(); ++i) {
    const Point& a = p[i];
    const Point& b = p[(i + 1) % p.size()];
    n.x += a.y * b.z - a.z * b.y;
    n.y += a.z * b.x - a.x * b.z;
    n.z += a.x * b.y - a.y * b.x;
  }
  return 0.5 * length(n);
}

// The signed volume of a tetrahedron; the pieces of a decomposition
// are summed signed and the total taken absolute, so that a
// decomposition of a cell whose faces are not planar still adds up.
double tet_volume(const Point& a, const Point& b, const Point& c,
                  const Point& d) {
  return dot(minus(b, a), cross(minus(c, a), minus(d, a))) / 6.0;
}

double sum_tets(const std::vector<Point>& p,
                const std::vector<std::array<int, 4>>& tets) {
  double v = 0.0;
  for (const std::array<int, 4>& t : tets) {
    v += tet_volume(p[static_cast<std::size_t>(t[0])],
                    p[static_cast<std::size_t>(t[1])],
                    p[static_cast<std::size_t>(t[2])],
                    p[static_cast<std::size_t>(t[3])]);
  }
  return std::fabs(v);
}

const char* cell_type_name(std::uint8_t code) {
  switch (code) {
    case 1: return "vertex";
    case 3: return "line";
    case 5: return "triangle";
    case 7: return "polygon";
    case 9: return "quadrilateral";
    case 10: return "tetrahedron";
    case 12: return "hexahedron";
    case 13: return "wedge";
    case 14: return "pyramid";
    case 21: return "quadratic line";
    case 22: return "quadratic triangle";
    case 23: return "quadratic quadrilateral";
    case 24: return "quadratic tetrahedron";
    case 25: return "quadratic hexahedron";
    case 26: return "quadratic wedge";
    case 27: return "quadratic pyramid";
    default: return "an unknown cell type";
  }
}

}  // namespace

int cell_dimension(std::uint8_t cell_type) {
  switch (cell_type) {
    case 3: return 1;                       // line
    case 5: case 7: case 9: return 2;       // triangle, polygon, quad
    case 10: case 12: case 13: case 14: return 3;
    default: return 0;                      // not measured here
  }
}

double cell_measure(std::uint8_t cell_type,
                    const std::vector<std::int64_t>& nodes,
                    const std::vector<double>& coordinates,
                    std::int64_t components) {
  const int dimension = cell_dimension(cell_type);
  if (dimension == 0) {
    throw Error("", std::string("mestra::cell_measure: a ") +
                        cell_type_name(cell_type) + " (cell type " +
                        internal::format_i64(cell_type) +
                        ") has no measure this version computes; section "
                        "3 asks for the measure of the linear cell types "
                        "and this build stops there");
  }
  if (components < dimension) {
    throw Error("", std::string("mestra::cell_measure: a ") +
                        cell_type_name(cell_type) + " is " +
                        internal::format_i64(dimension) +
                        "-dimensional and the coordinates have " +
                        internal::format_i64(components) + " component(s)");
  }
  std::vector<Point> p;
  p.reserve(nodes.size());
  for (const std::int64_t n : nodes) {
    p.push_back(point_at(coordinates, components, n));
  }
  switch (cell_type) {
    case 3:
      return length(minus(p[1], p[0]));
    case 5:
    case 7:
    case 9:
      return polygon_area(p);
    case 10:
      return std::fabs(tet_volume(p[0], p[1], p[2], p[3]));
    case 12:
      // The six tetrahedra that share the main diagonal 0-6 of a
      // hexahedron in the node order of section 20.
      return sum_tets(p, {{{0, 1, 2, 6}}, {{0, 2, 3, 6}}, {{0, 3, 7, 6}},
                          {{0, 7, 4, 6}}, {{0, 4, 5, 6}}, {{0, 5, 1, 6}}});
    case 13:
      return sum_tets(p, {{{0, 1, 2, 3}}, {{1, 2, 3, 4}}, {{2, 3, 4, 5}}});
    case 14:
      return sum_tets(p, {{{0, 1, 2, 4}}, {{0, 2, 3, 4}}});
    default:
      break;
  }
  throw Error("", std::string("mestra::cell_measure: no measure for ") +
                      cell_type_name(cell_type));
}

ArraySlot& compute_weights(Support& s, Location where,
                           const WeightOptions& options) {
  const std::string path = "/supports/" + s.name +
                           (where == Location::Node ? "/node_arrays/"
                                                    : "/cell_arrays/") +
                           options.name;
  if (!s.coordinates.has_value()) {
    refuse(path,
           "the support carries no coordinates, and a measure is "
           "computed from coordinates and connectivity");
  }
  const ArraySlot& xyz = *s.coordinates;
  if (xyz.is_callable() || xyz.data.f64.empty()) {
    refuse(path,
           "the support's coordinates hold no values, and a measure is "
           "computed from coordinates and connectivity");
  }
  const std::int64_t components = xyz.components;
  const std::size_t nodes = static_cast<std::size_t>(s.n_nodes);
  const std::size_t cells = static_cast<std::size_t>(s.n_cells);
  const std::size_t per_instance = nodes * static_cast<std::size_t>(
                                               components < 1 ? 1 : components);
  const std::size_t instances =
      per_instance == 0 ? 1 : std::max<std::size_t>(
                                  1, xyz.data.f64.size() / per_instance);

  if (s.kind == "none") {
    refuse(path, "a support of kind none has neither nodes nor cells");
  }
  if (where == Location::Cell && s.kind != "mesh") {
    refuse(path, "a support of kind " + s.kind +
                     " has no cells; ask for the node weights instead");
  }

  // The dimension of the measure, and one message rather than many
  // when the mesh mixes dimensions.
  int dimension = 0;
  if (s.kind == "mesh") {
    for (std::size_t c = 0; c < cells && c < s.cell_types.size(); ++c) {
      const int d = cell_dimension(s.cell_types[c]);
      if (d == 0) {
        refuse(path, std::string("cell ") + internal::format_i64(
                         static_cast<std::int64_t>(c)) + " is a " +
                         cell_type_name(s.cell_types[c]) +
                         " and this version measures the linear cell "
                         "types only");
      }
      if (dimension == 0) {
        dimension = d;
      } else if (dimension != d) {
        refuse(path,
               "the mesh mixes cells of " +
                   internal::format_i64(dimension) + " and " +
                   internal::format_i64(d) +
                   " dimensions, so its cells have no one measure");
      }
    }
    if (dimension == 0) dimension = 1;
  } else {
    dimension = 1;                       // an axis support is a line
  }

  const std::size_t count = where == Location::Node ? nodes : cells;
  std::vector<double> weight(instances * count, 0.0);

  for (std::size_t instance = 0; instance < instances; ++instance) {
    std::vector<double> block;
    const std::size_t from = instance * per_instance;
    if (from + per_instance <= xyz.data.f64.size()) {
      block.assign(xyz.data.f64.begin() + static_cast<std::ptrdiff_t>(from),
                   xyz.data.f64.begin() +
                       static_cast<std::ptrdiff_t>(from + per_instance));
    } else {
      block = xyz.data.f64;
    }
    double* out = weight.data() + instance * count;

    if (s.kind == "axis") {
      // No cells: the node's share is half of each segment it touches,
      // which is the trapezoid rule and the lumped measure of a line.
      // An axis support's coordinates have one component (section 6),
      // but the stride is taken from the slot rather than assumed.
      const std::size_t stride =
          components < 1 ? 1 : static_cast<std::size_t>(components);
      auto position = [&block, stride](std::size_t n) {
        const std::size_t at = n * stride;
        return at < block.size() ? block[at] : 0.0;
      };
      for (std::size_t n = 0; n < nodes; ++n) {
        double w = 0.0;
        if (n + 1 < nodes) {
          w += 0.5 * std::fabs(position(n + 1) - position(n));
        }
        if (n > 0) w += 0.5 * std::fabs(position(n) - position(n - 1));
        out[n] = w;
      }
      continue;
    }

    for (std::size_t c = 0; c < cells; ++c) {
      if (c + 1 >= s.cell_offsets.size()) break;
      const std::size_t begin =
          static_cast<std::size_t>(s.cell_offsets[c]);
      const std::size_t end =
          static_cast<std::size_t>(s.cell_offsets[c + 1]);
      if (end > s.cell_connectivity.size() || end < begin) break;
      const std::vector<std::int64_t> cell(
          s.cell_connectivity.begin() + static_cast<std::ptrdiff_t>(begin),
          s.cell_connectivity.begin() + static_cast<std::ptrdiff_t>(end));
      const double measure =
          cell_measure(s.cell_types[c], cell, block, components);
      if (where == Location::Cell) {
        out[c] = measure;
      } else if (!cell.empty()) {
        // Lumped: the cell's measure shared equally among its nodes.
        const double share = measure / static_cast<double>(cell.size());
        for (const std::int64_t n : cell) {
          if (n >= 0 && static_cast<std::size_t>(n) < nodes) {
            out[static_cast<std::size_t>(n)] += share;
          }
        }
      }
    }
  }

  ArraySlot a;
  a.name = options.name;
  a.role = "weight";
  a.varies = xyz.varies;
  a.components = 1;
  a.source = "data";
  a.recomputed = true;
  a.location = where;
  if (xyz.units.has_value()) {
    a.units = internal::units_power(*xyz.units, dimension);
  }
  a.data.dtype = DType::Float64;
  if (a.varies != "none") {
    a.data.dims.push_back(a.varies);
    a.data.shape.push_back(instances);
  }
  a.data.dims.push_back(where == Location::Node ? "node" : "cell");
  a.data.shape.push_back(count);
  a.data.dims.push_back("component");
  a.data.shape.push_back(1);
  a.data.f64 = std::move(weight);

  std::vector<ArraySlot>& into =
      where == Location::Node ? s.node_arrays : s.cell_arrays;
  for (std::size_t i = 0; i < into.size(); ++i) {
    if (into[i].name == a.name) {
      into[i] = std::move(a);
      return into[i];
    }
  }
  auto at = std::lower_bound(
      into.begin(), into.end(), a,
      [](const ArraySlot& x, const ArraySlot& y) {
        return bytes_less(x.name, y.name);
      });
  return *into.insert(at, std::move(a));
}

}  // namespace mestra
